File: typescript/packages/extensions/src/sign-in-with-x/message.ts
/**
 * CAIP-122 message construction for SIWX extension
 *
 * Constructs the canonical message string for signing.
 * Routes to chain-specific formatters based on chainId namespace.
 */

import { formatSIWEMessage } from "./evm";
import { formatSIWSMessage } from "./solana";
import type { CompleteSIWxInfo } from "./client";

/**
 * Construct CAIP-122 compliant message string for signing.
 *
 * Routes to the appropriate chain-specific message formatter based on the
 * chainId namespace prefix:
 * - `eip155:*` → SIWE (EIP-4361) format via siwe library
 * - `solana:*` → SIWS format
 *
 * @param serverInfo - Server extension info with chain selected (includes chainId)
 * @param address - Client wallet address
 * @returns Message string ready for signing
 * @throws Error if chainId namespace is not supported
 *
 * @example
 * ```typescript
 * // EVM (Ethereum, Base, etc.)
 * const completeInfo = { ...extension.info, chainId: "eip155:8453", type: "eip191" };
 * const evmMessage = createSIWxMessage(completeInfo, "0x1234...");
 * ```
 */
export function createSIWxMessage(serverInfo: CompleteSIWxInfo, address: string): string {
  // Route by chain namespace
  if (serverInfo.chainId.startsWith("eip155:")) {
    return formatSIWEMessage(serverInfo, address);
  }

  if (serverInfo.chainId.startsWith("solana:")) {
    return formatSIWSMessage(serverInfo, address);
  }

  throw new Error(
    `Unsupported chain namespace: ${serverInfo.chainId}. ` +
      `Supported: eip155:* (EVM), solana:* (Solana)`,
  );
}

File: python/x402/extensions/sign_in_with_x/message.py
"""CAIP-122 message construction for SIWX extension."""

from __future__ import annotations

from typing import Any

from .evm import format_siwe_message
from .solana import format_siws_message


def create_siwx_message(server_info: Any, address: str) -> str:
    """Construct CAIP-122 compliant message string for signing."""
    chain_id = server_info.chain_id if hasattr(server_info, "chain_id") else server_info["chainId"]
    if chain_id.startswith("eip155:"):
        return format_siwe_message(server_info, address)
    if chain_id.startswith("solana:"):
        return format_siws_message(server_info, address)
    raise ValueError(
        f"Unsupported chain namespace: {chain_id}. Supported: eip155:* (EVM), solana:* (Solana)"
    )
